File: Others/Sort/Python/QuickSort.py
#from Sort import Sort

class QuickSort(object):

    def sort(self, arr):
        left = 0
        right = len(arr) - 1
        self.basicQuickSort(arr, left, right)

    def basicQuickSort(self, arr, left, right):
        if left >= right:
            return
        
        pivot = arr[left]
        low, high = left, right
        while low < high:
            while arr[low] <= pivot and low < right:
                low += 1
            while arr[high] >= pivot and high > left:                
                high -= 1
            if low >= high:
                break
            arr[low], arr[high] = arr[high], arr[low]

        arr[left], arr[high] = arr[high], arr[left]
        self.basicQuickSort(arr, left, high - 1)
        self.basicQuickSort(arr, high + 1, right)

    def threeWaySort(self, arr, left, right):
        if left >= right:
            return

        low, high = left, right
        mid = low + 1
        pivot = arr[left]

        while mid <= high:
            if arr[mid] < pivot:
                arr[mid], arr[low] = arr[low], arr[mid]
                low += 1
                mid += 1
            elif arr[mid] == pivot:
                mid += 1
            else:
                arr[mid], arr[high] = arr[high], arr[mid]
                high -= 1
        self.threeWaySort(arr, left, low - 1)
        self.threeWaySort(arr, high + 1, right)
    
if __name__ == "__main__":
    arr = [1, 2, 3, 2, 4, 7, 6, 5, 7, 6, 7, 5, 3]
    print(arr)
    qs = QuickSort()
    qs.sort(arr)
    print(arr)

File: Others/Sort/C++/QuickSort.cpp
#include <iostream>
#include <vector>

using namespace std;

class QuickSort {
private:
    void swap(vector<int> &arr, int a, int b) {
        int tmp = arr[a];
        arr[a] = arr[b];
        arr[b] = tmp;
    }
    int partition(vector<int> &arr, int left, int right) {
        int pivot = arr[left];
        int p = left;
        left++;

        while (true) {
            while (arr[left] < pivot) {
                left++;
            }
            while (arr[right] > pivot) {
                right--;
            }

            if (left >= right) {
                break;
            }

            swap(arr, left, right);
        }
        swap(arr, p, right);
        return right;
    }
    void quickSort(vector<int> &arr, int left, int right) {
        if (right <= left) {
            return;
        }
        int p = partition(arr, left, right);
        quickSort(arr, left, p - 1);
        quickSort(arr, p + 1, right);
    }
public:
    void sort(vector<int> &arr) {
        quickSort(arr, 0, arr.size() - 1);
    }

    void printArr(vector<int>&arr) {
        for (int i = 0; i < arr.size(); i++) {
            cout << arr[i] << " ";
        }
        cout << endl;
    }
};

int main(int argc, char *argv[]) {
    QuickSort qs;
    vector <int> arr({2, 3, 1, 5, 7, 9, 0, 10, 4, 5});
    qs.sort(arr);
    qs.printArr(arr);
}